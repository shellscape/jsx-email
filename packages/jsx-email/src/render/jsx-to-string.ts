/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';
import hash from 'hash-it';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

const promiseMap = new Map();

const wrapPromise = <TPromise extends Promise<any>>(promise: TPromise) => {
  let status = 'pending';
  let result: Awaited<TPromise>;
  const suspender = promise.then(
    (r) => {
      status = 'success';
      result = r;
    },
    (e) => {
      status = 'error';
      result = e;
    }
  );
  return {
    // eslint-disable-next-line consistent-return
    read() {
      if (status === 'pending') {
        throw suspender;
      } else if (status === 'error') {
        throw result;
      } else if (status === 'success') {
        return result;
      }
    }
  };
};

export const useData = <TData>(props: any, cb: () => Promise<TData>): TData => {
  const key = hash(props);
  let dataPromise;
  if (promiseMap.has(key)) {
    dataPromise = promiseMap.get(key);
  } else {
    dataPromise = wrapPromise(cb());
    promiseMap.set(key, dataPromise);
  }
  return dataPromise.read();
};

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error('Unsupported JSX element:', element as any);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }

    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
