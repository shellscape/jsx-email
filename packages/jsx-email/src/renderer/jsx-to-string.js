/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */
import chalk from 'chalk';
import { log } from '../log.js';
import { AttributeAliases, BooleanAttributes, EmptyObject, VoidElements } from './constants.js';
import { escapeString } from './escape-string.js';
import { stringifyStyles } from './stringify-styles.js';
import { isValidElement } from './is-valid-element.js';
const renderSuspense = async (children) => {
    try {
        const result = await jsxToString(children);
        return result;
    }
    catch (error) {
        if (error instanceof Promise) {
            await error;
            return renderSuspense(children);
        }
        throw error;
    }
};
/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element) {
    if (element == null) {
        return '';
    }
    if (typeof element === 'string') {
        return escapeString(element);
    }
    else if (typeof element === 'number') {
        return String(element);
    }
    else if (typeof element === 'boolean' || element == null) {
        return '';
    }
    else if (isIterable(element)) {
        let html = '';
        for (const child of element) {
            // eslint-disable-next-line no-await-in-loop
            html += await jsxToString(child);
        }
        return html;
    }
    // Handle Promise case
    if (element instanceof Promise) {
        const resolvedElement = await element;
        return jsxToString(resolvedElement);
    }
    if (!isValidElement(element)) {
        log.error(chalk `{red Unsupported JSX element}:`, element);
        throw new Error(`Unsupported JSX element`);
    }
    const { type } = element;
    const props = element.props || EmptyObject;
    if (typeof type === 'string') {
        let html = `<${type}`;
        let innerHTML = '';
        for (const prop of Object.keys(props)) {
            const value = props[prop];
            if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
                // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
            }
            else if (prop === 'class' || prop === 'className') {
                // This condition is here because it is the most common attribute
                // and short-circuiting results in a ~5% performance boost.
                html += value ? ` class="${escapeString(value)}"` : '';
            }
            else if (prop === 'style') {
                html += ` style="${stringifyStyles(value)}"`;
            }
            else if (prop === 'dangerouslySetInnerHTML') {
                // eslint-disable-next-line no-underscore-dangle
                innerHTML = value.__html;
            }
            else {
                const name = AttributeAliases[prop] || prop;
                if (BooleanAttributes.has(name)) {
                    html += value ? ` ${name}` : '';
                }
                else if (typeof value === 'string') {
                    html += ` ${name}="${escapeString(value)}"`;
                }
                else if (typeof value === 'number') {
                    html += ` ${name}="${String(value)}"`;
                }
                else if (typeof value === 'boolean') {
                    html += ` ${name}="${value}"`;
                }
            }
        }
        if (VoidElements.has(type)) {
            html += '/>';
        }
        else {
            html += '>';
            if (innerHTML) {
                html += innerHTML;
            }
            else {
                html += await jsxToString(props.children);
            }
            html += `</${type}>`;
        }
        return html;
    }
    else if (type) {
        if (typeof type === 'function') {
            const result = type(props);
            const renderedFC = await (result instanceof Promise
                ? result.then(jsxToString)
                : jsxToString(result));
            const sym = type.$$typeof;
            if (sym && Symbol.keyFor(sym) === 'react.provider') {
                type.context.pop();
            }
            return renderedFC;
        }
        else if (typeof type === 'symbol') {
            const key = Symbol.keyFor(type);
            // is this react fragment?
            if (key === 'react.fragment') {
                return jsxToString(props.children);
            }
            else if (key === 'react.suspense') {
                const suspenseResult = await renderSuspense(props.children);
                return suspenseResult;
            }
        }
        else if (isReactForwardRef(type)) {
            return jsxToString(type.render(props, props.ref));
        }
        else if (type.$$typeof) {
            const key = Symbol.keyFor(type.$$typeof);
            if (key === 'react.provider') {
                type.context.pop();
            }
            else if (key === 'react.context') {
                return jsxToString(props.children);
            }
        }
        log.error(chalk `{red Unsupported JSX element}:`, type);
        throw new Error(`Unsupported JSX element type ${JSON.stringify(type)}`);
    }
    return '';
}
function isIterable(node) {
    return typeof node === 'object' && node !== null && Symbol.iterator in node;
}
function isReactForwardRef(type) {
    return (typeof type === 'object' &&
        type !== null &&
        'render' in type &&
        '$$typeof' in type &&
        typeof type.$$typeof === 'symbol' &&
        Symbol.keyFor(type.$$typeof) === 'react.forward_ref');
}
//# sourceMappingURL=jsx-to-string.js.map