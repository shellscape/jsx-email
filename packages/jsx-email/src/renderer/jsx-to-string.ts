/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */
import chalk from 'chalk';
import type { FC, ReactNode } from 'react';

import { log } from '../log.js';

import { AttributeAliases, BooleanAttributes, EmptyObject, VoidElements } from './constants.js';
import { escapeString } from './escape-string.js';
import { stringifyStyles } from './stringify-styles.js';

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    log.error(chalk`{red Unsupported JSX element}:`, element);
    throw new Error(`Unsupported JSX element`);
  }

  const { type } = element;
  const props = element.props || EmptyObject;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = AttributeAliases[prop as keyof typeof AttributeAliases] || prop;

        if (BooleanAttributes.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VoidElements.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      const renderedFC = await jsxToString((type as FC)(props));
      const sym = (type as { $$typeof?: symbol }).$$typeof;

      if (sym && Symbol.keyFor(sym) === 'react.provider') {
        (type as any as { context: any[] }).context.pop();
      }

      return renderedFC;
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    } else if ((type as { $$typeof?: symbol }).$$typeof) {
      const key = Symbol.keyFor((type as { $$typeof: symbol }).$$typeof);
      if (key === 'react.provider') {
        (type as any as { context: any[] }).context.pop();
      } else if (key === 'react.context') {
        return jsxToString(props.children);
      }
    }

    log.error(chalk`{red Unsupported JSX element}:`, type);
    throw new Error(`Unsupported JSX element type ${JSON.stringify(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
