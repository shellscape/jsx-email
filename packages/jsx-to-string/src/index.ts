export * from './jsx-to-string';
