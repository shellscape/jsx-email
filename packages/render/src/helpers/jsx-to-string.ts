/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export function jsxToString(element: ReactNode): string {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      html += jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error(`Unsupported JSX element: ${String(element)}`);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      // is this react fragment?
      if (Symbol.keyFor(type) === 'react.fragment') {
        return jsxToString(props.children);
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }
    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
