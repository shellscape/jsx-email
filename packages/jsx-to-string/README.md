[npm]: https://img.shields.io/npm/v/@jsx-email/jsx-to-string
[npm-url]: https://www.npmjs.com/package/@jsx-email/jsx-to-string

[![npm][npm]][npm-url]
[![Join our Discord](https://img.shields.io/badge/join_our-Discord-5a64ea)](https://discord.gg/FywZN57mTg)
[![libera manifesto](https://img.shields.io/badge/libera-manifesto-lightgrey.svg)](https://liberamanifesto.com)

<div align="center">
	<img src="https://raw.githubusercontent.com/shellscape/jsx-email/main/assets/npm-header.svg" alt="JSX email"><br/><br/>
</div>

# @jsx-email/jsx-to-string

<div>
  <img src="https://raw.githubusercontent.com/shellscape/jsx-email/main/assets/brackets.svg" alt="JSX email" valign="sub">
  A package for rendering JSX components to HTML strings.
<div><br>

[`@jsx-email/jsx-to-string` Documentation](https://jsx.email/docs/core/jsx-to-string)

## Requirements

This package requires an [LTS](https://github.com/nodejs/Release) Node version (v18.0.0+) and React v18.2.0+.

## Install

```shell
pnpm add @jsx-email/jsx-to-string

# We recommend pnpm - https://pnpm.io
# But npm and yarn are supported
# npm add @jsx-email/jsx-to-string
# yarn add @jsx-email/jsx-to-string
```

## Usage

Convert React components into a HTML string.

```jsx
import { jsxToString } from '@jsx-email/jsx-to-string';

const html = jsxToString(<MyTemplate firstName="Jim" />);
```

Please see the [`@jsx-email/jsx-to-string` Documentation](https://jsx.email/docs/core/jsx-to-string) for props and additional usage information.

## License

[MIT License](./LICENSE.md)
