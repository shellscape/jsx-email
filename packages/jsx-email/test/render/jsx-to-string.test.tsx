/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';
// @ts-ignore
import React from 'react';

import { jsxToString } from '../../src/renderer/jsx-to-string.js';

const Template: FC<{
  firstName: string;
}> = ({ firstName }) => (
  <>
    <h1>Welcome, {firstName}!</h1>
    <img src="img/test.png" alt="test" />
    <p>Thanks for trying our product. We're thrilled to have you on board!</p>
  </>
);

describe('jsx-to-string', async () => {
  it('converts a React component into HTML', async () => {
    const actualOutput = await jsxToString(<Template firstName="Jim" />);
    expect(actualOutput).toMatchSnapshot();
  });

  describe('elements', async () => {
    it('renders simple elements', async () => {
      const result = await jsxToString(<div></div>);
      expect(result).toBe('<div></div>');
    });

    it('renders void elements', async () => {
      const result = await jsxToString(<br />);
      expect(result).toBe('<br/>');
    });

    it('does not render fragments', async () => {
      const result = await jsxToString(<></>);
      expect(result).toBe('');
    });

    it('invokes functional components', async () => {
      const Component = (props: { text: string }) => <span className="c">{props.text}</span>;
      const result = await jsxToString(<Component text="Hello World" />);
      expect(result).toBe('<span class="c">Hello World</span>');
    });

    it('does not render null', async () => {
      expect(await jsxToString(null)).toBe('');
    });

    it('does not render booleans', async () => {
      expect(await jsxToString(true)).toBe('');
      expect(await jsxToString(false)).toBe('');
    });
  });

  describe('properties', async () => {
    it('renders HTML attribute names and values', async () => {
      const result = await jsxToString(<form action="/submit" method="post" />);
      expect(result).toBe('<form action="/submit" method="post"></form>');
    });

    it('does not render null or undefined HTML attributes', async () => {
      const result = await jsxToString(
        // @ts-expect-error because null is not a valid value for the `itemType` attribute
        // eslint-disable-next-line no-undefined
        <div itemType={null} itemProp={undefined} className={undefined} />
      );
      expect(result).toBe('<div></div>');
    });

    it('ignores framework specific properties', async () => {
      const result = await jsxToString(<div key="item" children={[]} />);
      expect(result).toBe('<div></div>');
    });

    it('aliases conflicting JS keywords', async () => {
      const result = await jsxToString(<label className="label" htmlFor="id" />);
      expect(result).toBe('<label class="label" for="id"></label>');
    });

    it('lowercases camelCase attribute names', async () => {
      const result = await jsxToString(<input tabIndex={-1} defaultValue="Hello" />);
      expect(result).toBe('<input tabindex="-1" value="Hello"/>');
    });

    it('escapes HTML attribute values', async () => {
      const result = await jsxToString(<img alt='"Mac & Cheese"' />);
      expect(result).toBe('<img alt="&quot;Mac &amp; Cheese&quot;"/>');
    });

    describe('boolean attributes', async () => {
      it('does not append boolean attributes with a falsy value', async () => {
        // @ts-expect-error because `open` is a boolean attribute.
        const result = await jsxToString(<details hidden={false} open={0} />);
        expect(result).toBe('<details></details>');
      });

      it('appends boolean attributes with a truthy value', async () => {
        // @ts-expect-error because `open` is a boolean attribute.
        const result = await jsxToString(<details hidden={true} open={1} />);
        expect(result).toBe('<details hidden open></details>');
      });

      it('renders boolean values for enumerable attributes', async () => {
        const result = await jsxToString(<div contentEditable={true} spellCheck={false} />);
        expect(result).toBe('<div contenteditable="true" spellcheck="false"></div>');
      });
    });

    describe('styles', async () => {
      it('stringifies style attributes', async () => {
        const result = await jsxToString(<div style={{ margin: '1em 0', padding: '0.5em 1em' }} />);
        expect(result).toBe('<div style="margin:1em 0;padding:0.5em 1em"></div>');
      });

      it('hyphenates style properties', async () => {
        const result = await jsxToString(
          <div style={{ marginBottom: '1em', paddingTop: '0.5em' }} />
        );
        expect(result).toBe('<div style="margin-bottom:1em;padding-top:0.5em"></div>');
      });

      it('hyphenates vendor prefixed properties', async () => {
        const result = await jsxToString(<div style={{ MozHyphens: 'auto', msHyphens: 'auto' }} />);
        expect(result).toBe('<div style="-moz-hyphens:auto;-ms-hyphens:auto"></div>');
      });

      it('applies pixel units to number values', async () => {
        const result = await jsxToString(<div style={{ margin: 20, padding: 10 }} />);
        expect(result).toBe('<div style="margin:20px;padding:10px"></div>');
      });

      it('does not apply pixel values to unitless properties', async () => {
        const result = await jsxToString(<div style={{ flexShrink: 1, order: 2 }} />);
        expect(result).toBe('<div style="flex-shrink:1;order:2"></div>');
      });

      it('ignores null or undefined properties', async () => {
        const result = await jsxToString(
          // @ts-expect-error because `width`, height` and `border` are not valid style properties.
          // eslint-disable-next-line no-undefined
          <div style={{ border: 0, height: undefined, width: null }} />
        );
        expect(result).toBe('<div style="border:0"></div>');
      });
    });
  });

  describe('children', async () => {
    it('ignores empty children', async () => {
      const result = await jsxToString(
        <div>
          {/* eslint-disable-next-line no-undefined */}
          {undefined}
          {null}
        </div>
      );
      expect(result).toBe('<div></div>');
    });

    it('renders text children', async () => {
      const result = await jsxToString(<div>Hello World</div>);
      expect(result).toBe('<div>Hello World</div>');
    });

    it('renders numeric children', async () => {
      const result = await jsxToString(<div>{123}</div>);
      expect(result).toBe('<div>123</div>');
    });

    it('renders numeric children even if they are zero', async () => {
      const result = await jsxToString(<div>{0}</div>);
      expect(result).toBe('<div>0</div>');
    });

    it('does not render boolean children', async () => {
      const a = await jsxToString(<div>{true}</div>);
      expect(a).toBe('<div></div>');

      const b = await jsxToString(<div>{false}</div>);
      expect(b).toBe('<div></div>');
    });

    it('escapes text children', async () => {
      const result = await jsxToString(<div>{'"Mac & Cheese"'}</div>);
      expect(result).toBe('<div>&quot;Mac &amp; Cheese&quot;</div>');
    });

    it('renders multiple children', async () => {
      const result = await jsxToString(
        <div>
          <i>Hello</i> <i>World</i>!
        </div>
      );
      expect(result).toBe('<div><i>Hello</i> <i>World</i>!</div>');
    });

    it('renders nested children', async () => {
      const result = await jsxToString(
        <div>{[<i key={1}>Hello</i>, [' ', [<i key={2}>World</i>, '!']]]}</div>
      );
      expect(result).toBe('<div><i>Hello</i> <i>World</i>!</div>');
    });

    it('passes children to compositional components', async () => {
      const Parent = (props: { children: ReactNode }) => <ul>{props.children}</ul>;
      const result = await jsxToString(
        <Parent>
          <li>one</li>
          <li>two</li>
        </Parent>
      );

      expect(result).toBe('<ul><li>one</li><li>two</li></ul>');
    });

    it('it allows children provided as props', async () => {
      const Parent = (props: { children: ReactNode }) => <ul>{props.children}</ul>;
      const result = await jsxToString(
        <Parent children={[<li key={1}>one</li>, <li key={2}>two</li>]} />
      );

      expect(result).toBe('<ul><li>one</li><li>two</li></ul>');
    });

    it('favours children provided as arguments over props', async () => {
      const Parent = (props: { children: ReactNode }) => <ul>{props.children}</ul>;
      const result = await jsxToString(
        // @ts-expect-error because `children` is specified twice
        <Parent children={[<li>one</li>, <li>two</li>]}>
          <li>three</li>
        </Parent>
      );

      expect(result).toBe('<ul><li>three</li></ul>');
    });

    it('supports setting inner HTML', async () => {
      const result = await jsxToString(
        <div dangerouslySetInnerHTML={{ __html: '<i>Hello</i>' }} />
      );
      expect(result).toBe('<div><i>Hello</i></div>');
    });

    it('renders the children of fragments', async () => {
      const result = await jsxToString(
        <dl>
          <>
            <dt>Title</dt>
            <dd>Description</dd>
          </>
        </dl>
      );
      expect(result).toBe('<dl><dt>Title</dt><dd>Description</dd></dl>');
    });
  });
});
