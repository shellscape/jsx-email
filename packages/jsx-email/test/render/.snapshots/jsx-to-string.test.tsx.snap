// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx-to-string > converts a React component into HTML 1`] = `"<h1>Welcome, Jim!</h1><img src=\\"img/test.png\\" alt=\\"test\\"/><p>Thanks for trying our product. We&#x27;re thrilled to have you on board!</p>"`;
